tonic/src/transport/channel/mod.rs

```diff
@@ -103,18 +106,23 @@ impl Channel {
     ///
     /// This creates a [`Channel`] that will load balance accross all the
     /// provided endpoints.
-    pub fn balance_list(list: impl Iterator<Item = Endpoint>) -> Self {
-        let list = list.collect::<Vec<_>>();
-
-        let buffer_size = list
-            .iter()
-            .next()
-            .and_then(|e| e.buffer_size)
-            .unwrap_or(DEFAULT_BUFFER_SIZE);
+    pub fn balance_list(list: impl Iterator<Item = Endpoint> + Send + 'static) -> Self {
+        let list = list.map(|endpoint| Change::Insert(endpoint.uri.clone(), endpoint));
 
-        let discover = ServiceList::new(list);
+        Self::balance_channel(tokio::stream::iter(list))
+    }
 
-        Self::balance(discover, buffer_size)
+    /// Balance a list of [`Endpoint`]'s.
+    ///
+    /// This creates a [`Channel`] that will listen to a stream of change events and will add or remove provided endpoints.
+    pub fn balance_channel<K>(
+        changes: impl Stream<Item = Change<K, Endpoint>> + Unpin + Send + 'static,
```

Comment:
So for the sake of simplicity I think this should take a `size: usize` and return the `tx` end of the channel. For the `balance_list` fn we can just iterate through all the endpoints and push them into the channel.

Reply:
done